src/Simulation/Simulators/SparseSimulator/SparseSimulatorCS/Probes.cs

```diff
@@ -0,0 +1,360 @@
+    /// and returns it as a boolean array
+    /// </summary>
+    partial class Sample
+    {
+        private SparseSimulator simulator;
+        public class Native : Sample
+        {
+            public Native(IOperationFactory m) : base(m)
+            {
+                simulator = m as SparseSimulator;
+            }
+            public override Func<IQArray<Qubit>, IQArray<bool>> __Body__ => (__in__) =>
+            {
+                if (simulator != null)
+                {
+                    return new QArray<bool>(((SparseSimulatorProcessor)simulator.QuantumProcessor).Sample(__in__));
+                }
+                return new QArray<bool>( new bool[0]);
+            };
+        }
+    }
+
+    public partial class SparseSimulatorProcessor : QuantumProcessorBase {
+
+
+        // These two delegates exist so that C# can use StringCallback (which has a nice
+        // string object) and C++ can use DumpCallback (which can handle char*)
+        public delegate void StringCallback(string label, double real, double img);
+
+        private delegate void DumpCallback(StringBuilder label, double real, double img);
+        [DllImport(simulator_dll)]
+        private static extern void Dump_cpp(uint sim, uint max_qubits, DumpCallback callback);
+
+        // This gets called by the base class
+        public bool Dump(StringCallback callback,  int max_id, IQArray<Qubit> qubits = null)
+        {
+            DumpCallback _callback = (StringBuilder label_builder, double real, double imag) =>
+            {
+                callback(label_builder.ToString(), real, imag);
+            };
+
+            // It's expensive for the C++ simulator to separate two parts of a wavefunction,
+            // so if we just want the full state, it calls a different function     
+            if (qubits == null)
+            {
+                Dump_cpp(Id, (uint)max_id, _callback);
+                return true;
+            } else
+            {
+                return DumpQubits_cpp(Id, qubits.Count(), qubits.Select(x => x.Id).ToArray(), _callback) ;
+            }
+        }
+
+        [DllImport(simulator_dll)]
+        [return: MarshalAs(UnmanagedType.I1)] // necessary because C++ and C# represent bools differently
+        private static extern bool DumpQubits_cpp(uint sim, int length, int[] qubit_ids, DumpCallback callback);
+
+        [DllImport(simulator_dll)]
+        private static extern uint num_qubits_cpp(uint sim);
+
+        // Returns the amplitude of a specific state, given by a string representing its label
+        [DllImport(simulator_dll)]
+        private static extern void GetAmplitude_cpp(uint sim, uint label_length, char[] label, ref double real, ref double imag);
+
+        public Microsoft.Quantum.Math.Complex GetAmplitude(string label)
+        {
+            double real = 0;
+            double imag = 0;
+            GetAmplitude_cpp(Id, (uint)label.Length, label.ToCharArray(), ref real, ref imag);
```

Comment:
If you define the `GetAmplitude_cpp` method with `real` and `imag` as `out` instead of `ref`, this wrapper method gets a bit easier:

```csharp
GetAmplitude_cpp(Id, (uint)label.Length, length.ToCharArray(), out var real, out var image);
return new Microsoft.Quantum.Math.Complex((real, imag));
```